### PR TITLE
Expose feature mapping

### PR DESCRIPTION
**Pull request by danieldk**
Features are accessible using a dict-like interface through the
'features' property of tokens.

### CODE REVIEW

**Comment by danieldk** [comment]
@twuebi is this good to go? I would like to branch of a 0.1.0 release, so that we have an initial version with basic functionality.

File: src/sentence.rs [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
 use conllx::graph::{Node, Sentence};
-use conllx::token::{Token, TokenBuilder};
+use conllx::token::{Features, Token, TokenBuilder};
 use pyo3::class::basic::PyObjectProtocol;
 use pyo3::class::iter::PyIterProtocol;
+use pyo3::class::mapping::PyMappingProtocol;
 use pyo3::class::sequence::PySequenceProtocol;
 use pyo3::exceptions;
 use pyo3::prelude::*;
+use pyo3::types::PyDict;
```
Comment (twuebi) [inline src/sentence.rs]: unused import

File: src/sentence.rs [hunk 2]
```diff
@@ -1,13 +1,16 @@
 use std::cell::{RefCell, RefMut};
+use std::collections::BTreeMap;
```
Comment (twuebi) [inline src/sentence.rs]: unused import

File: src/sentence.rs [hunk 3]
```diff
@@ -224,3 +236,103 @@ impl PyObjectProtocol for PyToken {
...
+                        .filter_map(|(f, v)| v.as_ref().map(|v| (f, v)))
+                        .map(|(f, v)| format!("\"{}\": \"{}\"", f, v))
+                        .collect::<Vec<_>>();
+                    fvals.join(", ")
+                }
+                None => String::new(),
+            },
+        };
+
+        Ok(format!("Features {{{}}})", dict_repr))
```
Comment (twuebi) [inline src/sentence.rs]: Superfluous `)`

**Comment by danieldk** [comment]
> It would be nice to have some access to iterate over the features (like `dict.items()`) + to have features as part of the `Token.__repr__` string.

I would like to implement `__iter__` and `__contains__`, but that currently does not seems to be possible with pyo3, because it does not have the specializations such as:

```
impl<'p, T> PyMappingContainsProtocolImpl for T where T: PyMappingContainsProtocol<'p> {
// ...
}
```

Leading to errors such as

```
245 | #[pyproto]
    | ^^^^^^^^^^
    |
    = note: conflicting implementation in crate `pyo3`:
            - impl<'p, T> pyo3::class::mapping::PyMappingContainsProtocolImpl for T
              where T: pyo3::class::mapping::PyMappingProtocol<'p>;
```

**Comment by danieldk** [comment]
Addressed your other comments.

**Comment by twuebi** [comment]
> > It would be nice to have some access to iterate over the features (like `dict.items()`) + to have features as part of the `Token.__repr__` string.
> 
> I would like to implement `__iter__` and `__contains__`, but that currently does not seems to be possible with pyo3, because it does not have the specializations such as:
> 
> ```
> impl<'p, T> PyMappingContainsProtocolImpl for T where T: PyMappingContainsProtocol<'p> {
> // ...
> }
> ```
> 
> Leading to errors such as
> 
> ```
> 245 | #[pyproto]
>     | ^^^^^^^^^^
>     |
>     = note: conflicting implementation in crate `pyo3`:
>             - impl<'p, T> pyo3::class::mapping::PyMappingContainsProtocolImpl for T
>               where T: pyo3::class::mapping::PyMappingProtocol<'p>;
> ```

Maybe for the time being a non-built-in `contains` method?

**Comment by danieldk** [comment]
Good idea, maybe we can work around the iterator issue as well by implementing a normal iteration protocol.

**Comment by danieldk** [comment]
I have added a `contains` method. I'll skip the iteration protocol for now, since that's quite a bit more work (due to `Rc<RefCell<...>>`'ed sentences).

**Pull request merged**